### PR TITLE
feat: Add channel encoder

### PR DESCRIPTION
**Pull request by kristw**
🏆 Enhancements

* Add `ChannelEncoder` which takes user-defined channel definition, complete and parse into an object with set of utility functions useful for rendering that channel. 

The "rendering" mention here is framework-independent. Chart component developer is required to figure out how to implement the rendering themselves. `ChannelEncoder` acts as a middle layer to standardize chart component consumer API, then translate the configuration into actionable functions.

```ts
const encoder = new ChannelEncoder({
  name: 'x',
  channelType: 'X',
  definition: {
    type: 'quantitative',
    field: 'speed',
    title: 'Speed',
    format: '.2f',
    scale: {
      type: 'linear',
      domain: [0, 100],
      range: [0, 1],
    }
  },
});

encoder.encodeValue(50); // 0.5
encoder.encodeDatum({ speed: 50 }); // 0.5
encoder.formatValue(10); // 10.00
encoder.formatDatum({ speed: 50 }); // 50.00
encoder.getTitle(); // 'Speed'
encoder.getDomain([{ speed: 10 }, { speed: 20 }]); // [10, 20]
encoder.scale // scale function (e.g. d3.scaleLinear, d3.scaleLog, depends on the definition)
```

* Add function `inferFieldType()` to infer field type from channel type and field name

```ts
inferFieldType('X', 'speed'); // 'quantitative'
inferFieldType('Color', 'brand'); // 'nominal'
inferFieldType('X', 'time'); // 'temporal'
```

### CODE REVIEW

File: packages/superset-ui-encodable/src/parsers/scale/createScaleFromScaleType.ts [hunk 1]
```diff
@@ -44,11 +44,11 @@ export default function createScaleFromScaleType<Output extends Value>(
     case ScaleType.THRESHOLD:
       return scaleThreshold<number | string | Date, Output>();
     case ScaleType.ORDINAL:
-      return scaleOrdinal<HasToString, Output>();
+      return scaleOrdinal<HasToString | null | undefined, Output>();
```
Comment (williaster) [inline packages/superset-ui-encodable/src/parsers/scale/createScaleFromScaleType.ts]: any reason not to use the `type CategoricalScaleInput = HasToString | null | undefined` from `/types/Scale`?

File: packages/superset-ui-encodable/src/fillers/completeChannelDef.ts [hunk 2]
```diff
@@ -1,27 +1,51 @@
...
+import { isFieldDef, isValueDef, isTypedFieldDef } from '../typeGuards/ChannelDef';
 import completeAxisConfig, { CompleteAxisConfig } from './completeAxisConfig';
 import completeScaleConfig, { CompleteScaleConfig } from './completeScaleConfig';
-import { Value } from '../types/VegaLite';
+import { Value, ValueDef, Type } from '../types/VegaLite';
+import inferFieldType from './inferFieldType';
 
-type CompleteChannelDef<Output extends Value = Value> = Omit<
-  ChannelDef,
+export interface CompleteValueDef<Output extends Value = Value> extends ValueDef<Output> {
```
Comment (williaster) [inline packages/superset-ui-encodable/src/fillers/completeChannelDef.ts]: is this a `type` or a `variable`? I don't quite understand it, just wanted to check it was intentional.

File: packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    this.definition = definition;
+
+    this.getValue = createGetterFromChannelDef(definition);
+    this.formatValue = createFormatterFromChannelDef(definition);
+
+    const scale = definition.scale && createScaleFromScaleConfig(definition.scale);
+    this.encodeValue = scale === false ? identity : (value: ChannelInput) => scale(value);
+    this.scale = scale;
+
+    this.encodeDatum = this.encodeDatum.bind(this);
```
Comment (williaster) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts]: can you do `arrow` functions for these instead?

File: packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  }: {
+    name: string;
+    channelType: ChannelType;
+    definition: Def;
+  }) {
+    this.name = name;
+    this.channelType = channelType;
+
+    this.originalDefinition = originalDefinition;
+    const definition = completeChannelDef(this.channelType, originalDefinition);
```
Comment (williaster) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts]: could just use `this.definition` but don't feel strongly.

File: packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    this.scale = scale;
+
+    this.encodeDatum = this.encodeDatum.bind(this);
+    this.formatDatum = this.formatDatum.bind(this);
+    this.getValueFromDatum = this.getValueFromDatum.bind(this);
+  }
+
+  encodeDatum(datum: PlainObject): Output | null | undefined;
+  // eslint-disable-next-line no-dupe-class-members
+  encodeDatum(datum: PlainObject, otherwise: Output): Output;
```
Comment (williaster) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts]: nice overloads 😎

File: packages/superset-ui-encodable/src/fillers/inferFieldType.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { ChannelType } from '../types/Channel';
+import { isXOrY } from '../typeGuards/Channel';
+import { Type } from '../types/VegaLite';
+
+const temporalFieldNames = new Set(['time', 'date', 'datetime', 'timestamp']);
```
Comment (williaster) [inline packages/superset-ui-encodable/src/fillers/inferFieldType.ts]: is this a robust way to infer temporal? useful proxy and not sure of anything better 🤔

File: packages/superset-ui-encodable/src/fillers/completeChannelDef.ts (hunk 2 above)
Comment (kristw) [reply packages/superset-ui-encodable/src/fillers/completeChannelDef.ts]: It is a `type`.

File: packages/superset-ui-encodable/src/fillers/inferFieldType.ts (hunk 6 above)
Comment (kristw) [reply packages/superset-ui-encodable/src/fillers/inferFieldType.ts]: I agree it is not robust and very cheap heuristics. Just offer a bit smarter inference than defaulting to `quantitative` all the time.

**Pull request merged**